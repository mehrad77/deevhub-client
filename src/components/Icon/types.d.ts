import { TColors } from '../../types';

export interface IStyledIcon {
  /* specifies the padding of the icon */
  padding?: number;
  /* specifies the fontSize of the icon */
  size?: number;
  /* specifies the color of the icon */
  color?: TColors;
}

export type TIconName =
  | 'down-o'
  | 'retweet-o'
  | 'star-o'
  | 'star-f'
  | 'loading';

export interface IIconProps extends IStyledIcon {
  /* specifies the type of the icon */
  type: TIconName;
  spin?: boolean;
}
