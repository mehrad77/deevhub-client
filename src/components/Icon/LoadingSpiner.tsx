import React from 'react';
import styled from 'styled-components';

const Wrapper = styled.div`
  position: relative;
  width: 26px;
  height: 26px;
  margin: auto;

  .wBall {
    position: absolute;
    width: 25px;
    height: 25px;
    opacity: 0;
    transform: rotate(225deg);
    animation: orbit 4.8425s infinite;
    .wInnerBall {
      position: absolute;
      width: 3px;
      height: 3px;
      background: rgb(0, 0, 0);
      left: 0px;
      top: 0px;
      border-radius: 3px;
    }
  }

  #wBall_1 {
  }

  #wBall_2 {
    animation-delay: 0.203s;
  }

  #wBall_3 {
    animation-delay: 0.4265s;
  }

  #wBall_4 {
    animation-delay: 0.6295s;
  }

  #wBall_5 {
    animation-delay: 0.843s;
  }

  @keyframes orbit {
    0% {
      opacity: 1;
      z-index: 99;
      transform: rotate(180deg);
      animation-timing-function: ease-out;
    }

    7% {
      opacity: 1;
      transform: rotate(300deg);
      animation-timing-function: linear;
      origin: 0%;
    }

    30% {
      opacity: 1;
      transform: rotate(410deg);
      animation-timing-function: ease-in-out;
      origin: 7%;
    }

    39% {
      opacity: 1;
      transform: rotate(645deg);
      animation-timing-function: linear;
      origin: 30%;
    }

    70% {
      opacity: 1;
      transform: rotate(770deg);
      animation-timing-function: ease-out;
      origin: 39%;
    }

    75% {
      opacity: 1;
      transform: rotate(900deg);
      animation-timing-function: ease-out;
      origin: 70%;
    }

    76% {
      opacity: 0;
      transform: rotate(900deg);
    }

    100% {
      opacity: 0;
      transform: rotate(900deg);
    }
  }
`;

const LoadingSpiner = () => {
  return (
    <Wrapper className="windows8">
      <div className="wBall" id="wBall_1">
        <div className="wInnerBall"></div>
      </div>
      <div className="wBall" id="wBall_2">
        <div className="wInnerBall"></div>
      </div>
      <div className="wBall" id="wBall_3">
        <div className="wInnerBall"></div>
      </div>
      <div className="wBall" id="wBall_4">
        <div className="wInnerBall"></div>
      </div>
      <div className="wBall" id="wBall_5">
        <div className="wInnerBall"></div>
      </div>
    </Wrapper>
  );
};

export default LoadingSpiner;
