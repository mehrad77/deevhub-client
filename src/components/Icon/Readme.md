# Icons

```js
<Icon size={26} type="down-o" padding={10} />
<Icon size={26} type="star-f" color="orange" padding={10}/>
<Icon size={26} type="star-o" color="blue" padding={10}/>
<Icon size={26} type="retweet-o" color="orange" padding={10}/>
```
