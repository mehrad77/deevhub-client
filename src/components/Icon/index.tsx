import React, { FC } from 'react';
import styled from 'styled-components';
import { FaRegStar, FaStar, FaRetweet, FaAngleDown } from 'react-icons/fa';
import LoadingSpiner from './LoadingSpiner';
import { prop, theme } from 'styled-tools';
import { IStyledIcon, TIconName, IIconProps } from './types';

const StyledIcon = styled.span<IStyledIcon>`
  padding: ${prop('padding', 0)}px;
  font-size: ${prop('size', 0)}px;
  color: ${({ color }) => (color ? theme('colors.' + color) : 'black')};
  /* TODO: rewrite this lineHeight */
  line-height: 0px;

  &.icon-spin {
    animation: icon-spin 2s infinite linear;
  }

  @keyframes icon-spin {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(359deg);
    }
  }
`;

const getIcon = (name: TIconName) => {
  switch (name) {
    case 'down-o':
      return <FaAngleDown />;
    case 'retweet-o':
      return <FaRetweet />;
    case 'star-f':
      return <FaStar />;
    case 'star-o':
      return <FaRegStar />;
    case 'loading':
      return <LoadingSpiner />;
    default:
      return <LoadingSpiner />;
  }
};

const Icon: FC<IIconProps> = ({ type, size, padding, spin, ...rest }) => {
  return (
    <StyledIcon
      size={size}
      padding={padding}
      className={spin ? 'icon-spin' : ''}
      {...rest}
    >
      {getIcon(type)}
    </StyledIcon>
  );
};

Icon.defaultProps = {
  size: 12,
  padding: 0,
};

export default Icon;
