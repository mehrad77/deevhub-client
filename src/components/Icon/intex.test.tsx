import React from 'react';
import { render } from 'enzyme';
import Icon from '.';

describe('Icon', () => {
  it('Renders orange star Icon', () => {
    const starIcon = render(<Icon size={26} type="star-f" color="orange" padding={10} />);
    expect(starIcon).toMatchSnapshot();
  });
});
