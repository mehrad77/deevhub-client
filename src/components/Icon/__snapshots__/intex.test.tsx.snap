// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Renders orange star Icon 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bwzfXH jlsGZX",
      "color": "orange",
      "size": "26",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "fill": "currentColor",
          "height": "1em",
          "stroke": "currentColor",
          "stroke-width": "0",
          "viewBox": "0 0 576 512",
          "width": "1em",
          "xmlns": "http://www.w3.org/2000/svg",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "d": "M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z",
            },
            "children": Array [],
            "name": "path",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "d": undefined,
            },
            "x-attribsPrefix": Object {
              "d": undefined,
            },
          },
        ],
        "name": "svg",
        "namespace": "http://www.w3.org/2000/svg",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "fill": undefined,
          "height": undefined,
          "stroke": undefined,
          "stroke-width": undefined,
          "viewBox": undefined,
          "width": undefined,
          "xmlns": "http://www.w3.org/2000/xmlns/",
        },
        "x-attribsPrefix": Object {
          "fill": undefined,
          "height": undefined,
          "stroke": undefined,
          "stroke-width": undefined,
          "viewBox": undefined,
          "width": undefined,
          "xmlns": "",
        },
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "color": undefined,
      "size": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "color": undefined,
      "size": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
